### PR TITLE
[SDL2] Run testautomation with CTest + run tests on MSVC

### PR DESCRIPTION
**Pull request by madebr**
Using dummy audio and dummy video driver, 

These tests fail on all tested platforms (Linux, Macos and Windows):
- [x] [mouse_createFreeCursor](https://github.com/libsdl-org/SDL/issues/8308)
- [x] [mouse_setCursor](https://github.com/libsdl-org/SDL/issues/8308)
- [x] [mouse_getCursor](https://github.com/libsdl-org/SDL/issues/8308)
- [x] mouse_getMouseFocus
- [x] [mouse_createFreeColorCursor](https://github.com/libsdl-org/SDL/issues/8308)
- [x] [video_createWindowVariousFlags](https://github.com/libsdl-org/SDL/issues/8309)
- [x] video_getWindowDisplayMode
- [x] video_setWindowGrab [ci@Windows](https://github.com/libsdl-org/SDL/actions/runs/6440955384/job/17490282188?pr=8313#step:6:8723) [ci@Macos](https://github.com/libsdl-org/SDL/actions/runs/6440955388/job/17490282462?pr=8313#step:11:8620) [ci@Linux](https://github.com/libsdl-org/SDL/actions/runs/6440955388/job/17490282284?pr=8313#step:11:9101)
- [X] SDL_xyzPrimarySelectionText (occasionaly) **segfaults** [ci@Linux](https://github.com/madebr/SDL/actions/runs/6342144428/job/17227287141#step:11:3287) [ci@Windows](https://github.com/libsdl-org/SDL/actions/runs/6440955384/job/17490282057?pr=8313#step:6:2794)  (Fixed in https://github.com/libsdl-org/SDL/commit/df6db90dd44984136d3724715fc2451d1b01fd23)

these tests fail on Windows (on ci):
- [ ] audio_initOpenCloseQuitAudio [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720397?pr=8313#step:11:2388) (also on Linux)
- [ ] audio_pauseUnpauseAudio [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720397?pr=8313#step:11:2687) (also on Linux)
- [ ] video_setWindowCenteredOnDisplay [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720397?pr=8313#step:11:10004)
- [x] keyboard_setTextInputRectNegative [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720397?pr=8313#step:11:3824)
- [x] pow_regularCases [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720397?pr=8313#step:11:4603)
- [x] atan2_bothZeroCases [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:5018)
- [x] atan2_yZeroCases [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:5028)
- [x] atan2_xZeroCases [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:5038)
- [x] atan2_bothInfCases [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:5048)
- [x] atan2_yInfCases [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:5058)
- [x] atan2_xInfCases [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:5068)
- [x] stdlib_snprintf [ci](https://github.com/madebr/SDL/actions/runs/6443143856/job/17494739498#step:11:7756)

these tests fail on Linux (on ci):

- [ ] audio_initQuitAudio [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720676?pr=8313#step:11:2402)
- [ ] audio_initOpenCloseQuitAudio [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720542?pr=8313#step:11:2388) (also on Windows)
- [ ] audio_pauseUnpauseAudio [ci@Linux](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279720542?pr=8313#step:11:2687) (also on Windows)

these tests fail on Macos (on ci):
- [x] keyboard_setTextInputRectNegative [ci](https://github.com/libsdl-org/SDL/actions/runs/6363845370/job/17279721067?pr=8313#step:11:3651)
- [x] timer_delayAndGetTicks (occasionally) (this is probably the same issue as SD3's https://github.com/libsdl-org/SDL/actions/runs/6440955388/job/17490282462?pr=8313#step:11:8172)

Watcom Windows failures have been moved to https://github.com/libsdl-org/SDL/issues/8396

### CODE REVIEW

**Comment by smcv** [comment]
> mouse_createFreeCursor, mouse_setCursor, mouse_getCursor, mouse_createFreeColorCursor

These are #8308.

> video_createWindowVariousFlags

This is #8309.

I see several of the others on Linux, but I haven't got round to reporting them with full details yet.

**Comment by Kontrabant** [comment]
The video tests should be passing with the dummy driver now.

**Comment by madebr** [comment]
Yup, great job!
I updated the list above.

**Comment by Kontrabant** [comment]
Huh, why is the grab test failing? It works locally, and `ShowWindow` should be sending the focus event, even with the dummy driver.

**Comment by madebr** [comment]
> Huh, why is the grab test failing? It works locally, and `ShowWindow` should be sending the focus event, even with the dummy driver.

I can reproduce a failing `video_getSetWindowGrab`, but only when running multiple tests.
After manually disabling tests and suites, I found out the failure happens when `events_pushPumpAndPollUserevent` is run before the `video_getSetWindowGrab` test.

**Comment by Kontrabant** [comment]
Yep, `SDL_PollEvent` is bailing out early due to a sentinel event, and even if that is worked around, for some reason, the other grab tests still fail.  The event test is putting the event queue into some state that is causing problems.

**Comment by madebr** [comment]
SDL2 needs https://github.com/libsdl-org/SDL/commit/3ccfd361cb4ff081f7696e5fbc4174f5d4aa1ca1

**Comment by madebr** [comment]
> e: Calling `SDL_PollEvent` at the end of the `events_pushPumpAndPollUserevent` function until it returns 0 seems to fix things, but that's probably just working around some internal issue that should be investigated.
> 
> e2: Once `SDL_PollEvent` is called, it needs to be called repeatedly until it hits the sentinel, so it looks like we just have to drain the event queue after the user event test.

I applied 3ccfd361cb4ff081f7696e5fbc4174f5d4aa1ca1, but `video_setWindowGrab` keeps failing.

**Comment by Kontrabant** [comment]
a486eec is needed as well, as calling `SDL_Quit` shuts down the event system and it's never restarted.

**Comment by Kontrabant** [comment]
I think the problematic `audio_initQuitAudio` test can be ripped out of both SDL2 and SDL3. In SDL2 it's error-prone as it tries to initialize every built-in audio backend, which will obviously fail if the test system doesn't support one of them (e.g. OSS on most Linux systems), and in SDL3 it can never fail due to the audio subsystem refcount never going to zero, so the test is basically a no-op there.  On top of that, if the specified or default audio backend isn't supported or fails to initialize, the audio subsystem initialization in the common test framework will fail at startup, causing the test suite to exit, so this test will never even be reached.

`audio_initOpenCloseQuitAudio` suffers from a similar issue, and can probably be simplified to just test the functionality of the currently initialized backend.

**Comment by madebr** [comment]
> I think the problematic `audio_initQuitAudio` test can be ripped out of both SDL2 and SDL3. In SDL2 it's error-prone as it tries to initialize every built-in audio backend, which will obviously fail if the test system doesn't support one of them (e.g. OSS on most Linux systems), and in SDL3 it can never fail due to the audio subsystem refcount never going to zero, so the test is basically a no-op there. On top of that, if the specified or default audio backend isn't supported or fails to initialize, the audio subsystem initialization in the common test framework will fail at startup, causing the test suite to exit, so this test will never even be reached.

I do some shenanigans in https://github.com/libsdl-org/SDL/pull/8347 to unitialize the audio backend.
I indeed see the OSS backend always failing on my system.

**Comment by madebr** [comment]
There are 4 failing tests remaining on Linux/Windows that I would like someone else to look after:
- audio_initOpenCloseQuitAudio
- audio_pauseUnpauseAudio
- audio_initQuitAudio
- video_setWindowCenteredOnDisplay

@sezero 
There are also test failures with watcom

**Comment by Kontrabant** [comment]
I'm working on the audio tests now.

On Linux, we're just going to have to special-case OSS failure, as it is always built due to omnipresent headers, but actual support is extremely rare these days.

The other major issue is that some APIs will initialize correctly, but fail to open an output device if there is no underlying audio hardware, so we're going to have to check for output devices and only try to run the device open/playback tests if there is some underlying support on the system.

**Comment by sezero** [comment]
> @sezero There are also test failures with watcom

Seeing similar failures compared to mingw/x86 builds but some additional
ones such as return code failures from SDL_OpenAudio.  Haven't look very
closely, though.  However, I remember having tested a watcom win32 build
in real life, e.g. with quakespasm, and it had worked.  The last person
that messed with watcom win32 builds was @ccawley2011, maybe he has some
ideas about the additional failures not seen in mingw or msvc builds..

**Comment by madebr** [comment]
> > @sezero There are also test failures with watcom
> 
> Seeing similar failures compared to mingw/x86 builds but some additional ones such as return code failures from SDL_OpenAudio. Haven't look very closely, though. However, I remember having tested a watcom win32 build in real life, e.g. with quakespasm, and it had worked.

The mingw stdlib test failures had to do with SDL2 using mingw libc instead of the builtin uclibc. Perhaps watcom is also using its own runtime?
https://github.com/libsdl-org/SDL/blob/df6db90dd44984136d3724715fc2451d1b01fd23/CMakeLists.txt#L240-L243

**Comment by sezero** [comment]
> The mingw stdlib test failures had to do with SDL2 using mingw libc instead of the builtin uclibc. Perhaps watcom is also using its own runtime?

Watcom uses its own C runtime, yes, but it's not shared,i.e.: watcom-built SDL2 dll and the exe linking to that SDL2 dll both link to watcom's clib statically. (They can be made to link to watcom's clib dll, but it's not what our current makefile does.) Can that lead to any of the failures? Maybe, haven't thought about it that way.

File: test/testautomation_audio.c [hunk 1]
```diff
@@ -100,6 +100,12 @@ int audio_initQuitAudio()
 
         /* Call Init */
         result = SDL_AudioInit(audioDriver);
+#if defined(__linux__)
+        /* OSS is rarely seen on modern Linux distros, so failure is generally expected. */
+        if (result != 0 && SDL_strcmp(audioDriver, "dsp") == 0) {
```
Comment (madebr) [inline test/testautomation_audio.c:105]: Are we ok with not allowing testing the dsp driver exclusively? e.g.
```
SDL_AUDIO_DRIVER=dsp test/testautomation
```

**Comment by madebr** [comment]
> Can that lead to any of the failures? Maybe, haven't thought about it that way.

The 32-bit mingw test failures were (mostly) caused by mingw's runtime returning higher precision (80 bit?) floats, causing comparison tests to fail. It's possible the watcom runtime is less precise and/or has different behavior on bad input.

File: test/testautomation_audio.c (hunk 1 above)
Comment (Kontrabant) [reply test/testautomation_audio.c]: Ah, sorry, that was supposed to be pushed to my own repo for testing.

But yes, we probably should explicitly allow it.  The tests which change the audio driver hint should save and restore the original hint so this is possible.

**Comment by sezero** [comment]
> > Can that lead to any of the failures? Maybe, haven't thought about it that way.
> 
> The 32-bit mingw test failures were (mostly) caused by mingw's runtime returning higher precision (80 bit?) floats, causing comparison tests to fail. It's possible the watcom runtime is less precise and/or has different behavior on bad input.

Possible. My humble suggestion would be leaving testautomation out of interactive tests for watcom win32, at least for the time being..

**Comment by madebr** [comment]
> Possible. My humble suggestion would be leaving testautomation out of interactive tests for watcom win32, at least for the time being..

Yeah sure. Let's create a issue ticket so we remember to come back at a later time.

**Comment by Kontrabant** [comment]
Green checks are pretty :)

I don't particularly like having to skip the Jack driver in testing, but it just doesn't want to work.  Testing it under real Jack always fails at the point of initialization, and, while it will initialize via Pipewire compatibility layer, trying to open the device it always exposes will fail if no underlying output is present.

As for nas and sndio, no clue.  I can't get them to work, and I doubt anyone uses or particularly cares about them as they were ripped out of SDL3 with no protest.  I'm fine with leaving those at "patches welcome" if someone really wants them to pass automated testing.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -55,10 +55,23 @@ jobs:
...
       if: matrix.platform.os == 'ubuntu-22.04'
       run: |
-        sudo apt-get install libpipewire-0.3-dev libdecor-0-dev
+        sudo apt-get install libpipewire-0.3-dev libdecor-0-dev pipewire-media-session- wireplumber \
+        pipewire-audio-client-libraries
+        systemctl --user --now enable wireplumber.service
+        systemctl --user --now enable pipewire.service
+        systemctl --user --now enable pipewire.socket
+        systemctl --user --now enable pipewire-pulse.service
+        systemctl --user --now enable pipewire-pulse.socket
```
Comment (madebr) [inline .github/workflows/main.yml:74]: Are these needed?
CTest runs the tests with `SDL_AUDIODRIVER=dummy SDL_VIDEODRIVER=dummy`,

Comment (Kontrabant) [reply .github/workflows/main.yml:74]: Some of the audio tests (the ones which were failing) check all of the drivers, regardless of which one is selected. We either have to start the services they expect, or drop the tests.

**Comment by slouken** [comment]
> Green checks are pretty :)
> 
> I don't particularly like having to skip the Jack driver in testing, but it just doesn't want to work. Testing it under real Jack always fails at the point of initialization, and, while it will initialize via the Pipewire compatibility layer, trying to open the device it always exposes will fail if no underlying output is present.
> 
> As for nas and sndio, no clue. I can't get them to work, and I doubt anyone uses or particularly cares about them as they were ripped out of SDL3 with no protest. I'm fine with skipping them and leaving those at "patches welcome" if someone really wants them to pass automated testing.
> 
> In any case, you can still force these drivers to run the tests by setting `SDL_AUDIODRIVER`.

We probably shouldn't try to initialize all the audio drivers in CI. The NAS audio driver can probably be removed, I don't know of anyone still using it. The sndio driver I believe was for OpenBSD, we should check to see whether it's still the preferred audio interface there.

**Comment by Kontrabant** [comment]
The tests are now redesigned to only use the old behavior of looping over all drivers if `SDL_AUDIODRIVER` is undefined. If a specific driver was requested, as is the case in CI, it will only test the requested driver.

Noted that sndio is the OpenBSD API, and that it still needs to be skipped on Linux (unless specifically requested), as actual support is likely as rare as OSS, if not more so.  If we want to tear out NAS, we should probably do so in a separate pull, just in case someone is actually using it in SDL2 and wants to raise an objection. It's already gone in SDL3, so it's a solved problem there.

**Comment by slouken** [comment]
Oh, yeah, we shouldn't remove NAS in SDL2, that's an SDL3 thing. We can definitely leave it out of CI tests though. :)

**Comment by madebr** [comment]
The segmentation fault of https://github.com/madebr/SDL/actions/runs/6908412450/job/18797644010 looks weird.
`testatomic.exe` segfaults, while `testautomation.exe` has just been started (~1s).
FYI, the tests run in parallel `-j2`.

**Comment by madebr** [comment]
Is this ok to merge before the 2.30 release?

**Comment by slouken** [comment]
Yes, that's fine.

**Comment by madebr** [comment]
Going to merge now before it gets stale again :)

**Pull request merged**